In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option('display.max_columns', 100)

## Extract information from columns descriptions using the text file `dataset dictionary.txt`

In [2]:
content = []
with open('dataset dictionary.txt', 'r') as f:
    for line in f.readlines():
        content.append(line)

In [3]:
content = [i.strip('\n').split(': ') for i in content]

In [4]:
content = {i[0]: i[1] for i in content} 

In [5]:
var_description_df = pd.DataFrame({'variable': content.keys(), 'description': content.values()}, index = np.arange(len(content)))

In [6]:
var_description_df.set_index('variable', inplace=True)

In [7]:
var_description_df.to_csv('variable_descriptions.csv', index=False)

In [8]:
var_description_df

description
variable                                                      
photoUrl                        The URL of the player's photo.
LongName                          The full name of the player.
playerUrl          The URL of the player's page on sofifa.com.
Nationality                     The nationality of the player.
Positions                   The positions the player can play.
...                                                        ...
PAS          This refers to the player's passing ability. I...
DRI          This refers to the player's dribbling ability....
DEF          This refers to the player's defensive ability....
PHY          This refers to the player's physicality or str...
Hits         This refers to the number of times the player'...

[77 rows x 1 columns]

# Start of cleaning

In [9]:
df = pd.read_csv('fifa21 raw data v2.csv', low_memory=False)

In [10]:
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioni

Wow!! 77 columns. Let's see them more in depth:

In [11]:
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

## Missing values study

Let's study the missing values and their counts  
It's not the prettiest method but I can't remember now how to do it in a more elegant way...

In [12]:
s = pd.Series([df[i].isnull().sum() for i in df.columns], index=df.columns).sort_values(ascending=False)
s.name = 'missing values count'
s

Loan Date End       17966
Hits                 2595
GK Diving               0
Sliding Tackle          0
Standing Tackle         0
                    ...  
Heading Accuracy        0
Finishing               0
Crossing                0
Attacking               0
Reactions               0
Name: missing values count, Length: 77, dtype: int64

Study of the dtypes:

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

## Study of object variables

See if any object (mainly string) variable should be numeric

In [14]:
df.select_dtypes(object).columns

Index(['Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Club',
       'Contract', 'Positions', 'Height', 'Weight', 'Preferred Foot',
       'Best Position', 'Joined', 'Loan Date End', 'Value', 'Wage',
       'Release Clause', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'Hits'],
      dtype='object')

In [15]:
# just take 5 rows to check the values and see if they effectively must be integers or some need a type conversion
string_vars = df.select_dtypes(object).sample(5)
string_vars.T

18707  \
Name                                                    Wang Peng   
LongName                                                Peng Wang   
photoUrl         https://cdn.sofifa.com/players/245/029/21_60.png   
playerUrl       http://sofifa.com/player/245029/peng-wang/210006/   
Nationality                                              China PR   
Club                                     \n\n\n\nGuangzhou R&F FC   
Contract                                              2020 ~ 2023   
Positions                                                 CM, CDM   
Height                                                      176cm   
Weight                                                       69kg   
Preferred Foot                                              Right   
Best Position                                                 CAM   
Joined                                               Aug 10, 2020   
Loan Date End                                                 NaN   
Value                                                       €130K   
Wage                                                          €2K   
Release Clause                                              €105K   
W/F                                                           3 ★   
SM                                                             2★   
A/W                                                        Medium   
D/W                                                        Medium   
IR                                                            1 ★   
Hits                                                          NaN   

                                                            3969   \
Name                                                    A. Canelo   
LongName                                            Alexis Canelo   
photoUrl         https://cdn.sofifa.com/players/227/591/21_60.png   
playerUrl       http://sofifa.com/player/227591/alexis-canelo/...   
Nationality                                             Argentina   
Club                                     \n\n\n\nDeportivo Toluca   
Contract                                              2017 ~ 2021   
Positions                                              CF, ST, LW   
Height                                                      180cm   
Weight                                                       83kg   
Preferred Foot                                              Right   
Best Position                                                  CF   
Joined                                                Aug 9, 2017   
Loan Date End                                                 NaN   
Value                                                       €1.9M   
Wage                                                         €16K   
Release Clause                                              €4.1M   
W/F                                                           4 ★   
SM                                                             3★   
A/W                                                          High   
D/W                                                           Low   
IR                                                            1 ★   
Hits                                                            9   

                                                            542    \
Name                                                 Adnan Vidual   
LongName                           Adnan Claudiano Vidual Machado   
photoUrl         https://cdn.sofifa.com/players/230/216/21_60.png   
playerUrl       http://sofifa.com/player/230216/adnan-claudian...   
Nationality                                                Brazil   
Club                                            \n\n\n\nPalmeiras   
Contract                                              2019 ~ 2023   
Positions                                                      RB   
Height                                                      175cm   
Weight                                                  

## Proposed actions

Some conclusions after observing the data from some `object` type columns:  
1. Rename columns so it's easier to access them using dot notation (df.wage)
2. `Height` could be transformed to feet and converted to float
3. `Weight` can be converted to integer after removing lbs
4. `Foot` definitely must be converted to categorical
5. `Value` and `Release Clause` can be converted to float and remove euro sign (or maybe there's some dollar sign)
6. `W/F`, `SM` and `IR` must be converted to integer (or categorical and then ordered category)
7. `Joined` must be datetime dtype
8. `Hits` can be cleaned and transformed into integer (later converted to ordered categorical)
9. Set `a/w` and `d/w` to categorical and sort their category values
10. Study `contract` column
11. Clean `club` values
12. Reestructuration and/or removal of columns after all processing

### 1 Rename columns

In [16]:
columns_backup = df.columns.to_list()
df.columns = [i.replace(' & ', '_').replace(' ', '_').lower() for i in df.columns]

In [17]:
df.columns

Index(['id', 'name', 'longname', 'photourl', 'playerurl', 'nationality', 'age',
       '↓ova', 'pot', 'club', 'contract', 'positions', 'height', 'weight',
       'preferred_foot', 'bov', 'best_position', 'joined', 'loan_date_end',
       'value', 'wage', 'release_clause', 'attacking', 'crossing', 'finishing',
       'heading_accuracy', 'short_passing', 'volleys', 'skill', 'dribbling',
       'curve', 'fk_accuracy', 'long_passing', 'ball_control', 'movement',
       'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance',
       'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'mentality', 'aggression', 'interceptions', 'positioning', 'vision',
       'penalties', 'composure', 'defending', 'marking', 'standing_tackle',
       'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats',
       'base_stats', 'w/f', 'sm', 'a/w', 'd/w', 'ir', 'pac', 'sho', 'pas',
       'dri', 'def', 

### 2. `Height` could be transformed to feet and converted to float

In [18]:
height = df.height
height.sample(10)

5553     170cm
18138    179cm
16741    184cm
14664    183cm
1723     178cm
10941    178cm
14386    187cm
2082     189cm
572      176cm
18834    179cm
Name: height, dtype: object

All heights seem to be in cm. Let's see if there are some feet or inches

In [19]:
# number of row with cm in the value
height.str.contains('ft').sum()

0

In [20]:
# number of row with cm in the value
height.str.contains('inch').sum()

0

In [21]:
# number of row with cm in the value
height.str.contains('feet').sum()

0

All of the heights (which are not null) are in cm?

In [22]:
df.height[df.height.notnull()].str.endswith('cm').all()

False

Let's check what other units there are.

In [23]:
notnull_height = df.height.dropna()

In [24]:
notnull_height.str[3:].unique()

array(['cm', '"', '1"', '0"'], dtype=object)

The results show that there are some values in inches ( " )

In [25]:
notnull_height[notnull_height.str.endswith('"')]

793     6'2"
847     6'3"
848     6'5"
860    5'11"
861     6'4"
862     6'1"
863     6'0"
864     6'1"
865    5'11"
866     6'2"
867     6'0"
868     6'3"
869     6'0"
871    5'10"
872     5'9"
873    5'11"
874     5'6"
875     6'1"
876     6'0"
877     6'4"
878     5'7"
879     6'0"
880    5'11"
881    5'11"
883     6'0"
884    5'10"
885     6'3"
886     6'0"
887     5'4"
888    5'11"
889     5'9"
890     5'7"
891     6'0"
892    5'10"
893     6'2"
894     6'2"
895     6'3"
896     5'7"
897    5'10"
898     5'9"
Name: height, dtype: object

Definitely `height` needs some cleaning and uniform units for height. I will create `height_m` and `height_feet` columns.

In [26]:
df['height_m'] = df.height
df['height_feet'] = df.height

1 metre = 0.3048 feet
1 inch = 1/12 feet

In [27]:
def to_metres(value):
    # if unit is cm
    if value.endswith('cm'):
        value = int(value[:-2])
        # conversion from cm to metre
        value = value / 100
        
    # if value is in feet and inches
    else:
        value = value.strip('"').split("'")
        # conversion from feet and inches to just feet
        value = float(value[0]) + float(value[1])/12
        # conversion from feet to metre
        value *= 0.3048
        
    return value

In [28]:
def to_feet(value):
    # if unit is cm
    if value.endswith('cm'):
        value = int(value[:-2])
        # conversion from cm to feet
        value = value / 30.48
        
    # if value is in feet and inches
    else:
        value = value.strip('"').split("'")
        # conversion to feet
        value = float(value[0]) + float(value[1])/12
        
    return value

In [29]:
df.height_m = df.height_m.apply(to_metres)

In [30]:
df.height_feet = df.height_feet.apply(to_feet)

At the end, I will remove the unnecessary columns. By now, I will keep all the height columns.

### 3. `Weight` can be converted to integer after removing `kg`

In [31]:
weight = df.weight
weight.sample(5)

7931     74kg
6346     80kg
4068     87kg
14257    84kg
17550    65kg
Name: weight, dtype: object

#### Error: `ValueError: invalid literal for int() with base 10: '183l'`
Uncomment the cell below to see the whole error output.

In [32]:
#weight = weight.apply(lambda x : int(x[:-2]) )

It seems as if some weights are in lb and others in Kg

In [33]:
weight.str[-3:].unique()

array(['2kg', '3kg', '7kg', '0kg', '8kg', '1kg', '5kg', '9kg', '4kg',
       '6kg', 'lbs'], dtype=object)

In [34]:
def convert_weight(value):
    if value.endswith('kg'):
        return float(value[:-2])  # lbs to kg will throw float values
    else:
        value = value[:-3]
        value = float(value) / 2.204623  # 1 Kg = 2,204623 lb
        return value

In [35]:
weight = weight.apply(convert_weight)

In [36]:
weight

0        72.0
1        83.0
2        87.0
3        70.0
4        68.0
         ... 
18974    66.0
18975    65.0
18976    74.0
18977    69.0
18978    75.0
Name: weight, Length: 18979, dtype: float64

In [37]:
df.weight = weight

### 4. `Preferred Foot` definitely must be converted to categorical

In [38]:
df.preferred_foot

0         Left
1        Right
2        Right
3        Right
4        Right
         ...  
18974    Right
18975    Right
18976    Right
18977    Right
18978     Left
Name: preferred_foot, Length: 18979, dtype: object

In [39]:
foot =df.preferred_foot
foot.unique()

array(['Left', 'Right'], dtype=object)

In [40]:
df.preferred_foot = foot.astype('category')

In [41]:
df.dtypes['preferred_foot']

CategoricalDtype(categories=['Left', 'Right'], ordered=False)

### 5. `value`, `wage` and `release_clause` can be converted to float and remove euro sign (or maybe there's some dollar sign)

In [42]:
value = df.value
wage = df.wage
release_clause = df.release_clause

In [43]:
wage.sample(5)

18446    €500
3991      €9K
15622     €2K
17087     €1K
13365     €1K
Name: wage, dtype: object

In [44]:
value.sample(5)

13019      €1M
7507     €2.3M
17569    €400K
7422     €1.2M
12369    €575K
Name: value, dtype: object

In [45]:
release_clause.sample(5)

4836      €2.2M
12819     €799K
980      €24.2M
16164     €571K
16939     €289K
Name: release_clause, dtype: object

#### 5.1 Check if all values are in euros or there are in dollars

In [46]:
value.str[0].unique()

array(['€'], dtype=object)

In [47]:
wage.str[0].unique()

array(['€'], dtype=object)

In [48]:
release_clause.str[0].unique()

array(['€'], dtype=object)

#### 5.2 Remove euro sign

In [49]:
value = value.str[1:]
wage = wage.str[1:]
release_clause = release_clause.str[1:]

#### 5.3 Convert to million euros using a custom function

In [50]:
def to_millions(i):
	if i[-1] == 'K':
		return float(i[:-1]) / 1000
	elif i[-1] == 'M':
		return float(i[:-1])
	else:
		return float(i[:-1]) / 1E6		

#### Error: `ValueError: could not convert string to float: ''`
Uncomment the cell below to see the whole error output.

In [51]:
#value.apply(to_millions)

It looks as if there are some empty values. Count the nan for every variable

In [52]:
v = value.str[:-1]
(v == '').sum()

248

In [53]:
w = wage.str[:-1]
(w == '').sum()

237

In [54]:
r = release_clause.str[:-1]
(r == '').sum()

1261

Last count of nan values was really high. What is the percentage of nan in the variable?

In [55]:
nan_count = (r == '').sum()
print(f'The percentage of nan values in release_clause variable is: {round(nan_count / len(r) * 100, 2)}')

The percentage of nan values in release_clause variable is: 6.64


So the count was not so high in the end.

There are empty values: ''. I will use a slightly modified version of function `to_millions` that replaces the '' values for '0M'.

In [56]:
def to_millions(i):
	if i[:-1] == '':
		return np.nan
	elif i[-1] == 'K':
		return float(i[:-1]) / 1000
	elif i[-1] == 'M':
		return float(i[:-1])
	else:
		return float(i[:-1]) / 1E6

In [57]:
value = value.apply(to_millions)
wage = wage.apply(to_millions) * 52  # convert to annual wage
release_clause = release_clause.apply(to_millions)

#### 5.4 Impute nan values by the mean value

Remember that there were empty values, which are now nan values:

In [58]:
value.isna().sum()

248

I will impute nan values by the mean of every variable

In [59]:
value_mean = value[value.notnull()].mean()
value_mean

2.9029965832043136

In [60]:
wage_mean = wage[wage.notnull()].mean()
wage_mean

0.4730918631949631

In [61]:
release_clause_mean = release_clause[release_clause.notnull()].mean()
release_clause_mean

4.24499678293261

In [62]:
value = value.fillna(value_mean)
wage = wage.fillna(wage_mean)
release_clause = release_clause.fillna(release_clause_mean)

#### 5.5 Save the clean variables to the dataframe and rename column to `euros_M`

In [63]:
df.value = value
df.wage = wage
df.release_clause = release_clause

In [64]:
df.rename(columns = {'value' : 'value_euros_M',
                    'wage': 'annual_wage_euros_M',
                    'release_clause': 'release_clause_euros_M'},
          inplace='True')

### 6. `W/F`, `SM` and `IR` must be converted to integer (or categorical and then ordered category)

In [65]:
wf = df['w/f']
sm = df.sm
ir = df.ir

Inspect wf, sm and ir values

In [66]:
wf.unique()

array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

In [67]:
sm.unique()

array(['4★', '5★', '1★', '2★', '3★'], dtype=object)

In [68]:
ir.unique()

array(['5 ★', '3 ★', '4 ★', '2 ★', '1 ★'], dtype=object)

Custom function to process data:

In [69]:
def remove_star(i):
    return int(i[0])

In [70]:
wf.apply(remove_star)

0        4
1        4
2        3
3        5
4        5
        ..
18974    2
18975    2
18976    2
18977    3
18978    3
Name: w/f, Length: 18979, dtype: int64

In [71]:
wf = wf.apply(remove_star)
sm = sm.apply(remove_star)
ir = ir.apply(remove_star)

Now lets convert to category dtype

In [72]:
wf = wf.astype('category')
sm = sm.astype('category')
ir = ir.astype('category')

Check if categories are ordered:

#### Error: `TypeError: Unordered Categoricals can only compare equality or not`
Uncomment the cell below to see the whole error output.

In [73]:
#ir[ir >=3]

Set ordered categories using `.cat.set_categories([ordered values], ordered=True)`

In [74]:
wf = wf.cat.set_categories([1,2,3,4,5], ordered=True)
sm = sm.cat.set_categories([1,2,3,4,5], ordered=True)
ir = ir.cat.set_categories([1,2,3,4,5], ordered=True)

Check if categories are ordered (again):

In [75]:
ir[ir >=3]

0        5
1        5
2        3
3        4
4        5
        ..
4988     3
5792     3
6286     3
8552     3
10101    3
Name: ir, Length: 332, dtype: category
Categories (5, int64): [1 < 2 < 3 < 4 < 5]

Finally, assign to the original dataframe

In [76]:
df['w/f'] = wf
df.sm = sm
df.ir = ir

### 7. `Joined` must be datetime dtype

In [77]:
joined = df.joined

In [78]:
pd.to_datetime(joined)

0       2004-07-01
1       2018-07-10
2       2014-07-16
3       2015-08-30
4       2017-08-03
           ...    
18974   2018-07-13
18975   2020-08-01
18976   2019-03-08
18977   2020-09-22
18978   2019-07-29
Name: joined, Length: 18979, dtype: datetime64[ns]

In [79]:
pd.to_datetime(joined).isna().sum()

0

Everything went right, so let's assign

In [80]:
df.joined = pd.to_datetime(joined)

### 8. `Hits` can be cleaned and transformed into integer

In [81]:
hits = df.hits
hits

0        771
1        562
2        150
3        207
4        595
        ... 
18974    NaN
18975    NaN
18976    NaN
18977    NaN
18978    NaN
Name: hits, Length: 18979, dtype: object

We see there are some 'NaN' values. Let's make a count of nulls

In [82]:
df.hits.isnull().sum()

2595

Let's try a dtype conversion and see what happens...

#### Error: `ValueError: invalid literal for int() with base 10: '1.6K'`
Uncomment the cell below to see the whole error output.

In [83]:
#df.hits.astype(int)

Looks like some of them have letters (K for thousands). They need a deeper transformation.

In [84]:
hits.str.contains('K').sum()

28

In [85]:
hits.str.lstrip()

0        771
1        562
2        150
3        207
4        595
        ... 
18974    NaN
18975    NaN
18976    NaN
18977    NaN
18978    NaN
Name: hits, Length: 18979, dtype: object

In [86]:
def clean_hits(i):
    i = i.lstrip()
    if i.endswith('K'):
        i = i[:-1]
        i = float(i) * 1000
        i = int(i)
    else:
        i = int(i)
    return i

#### Error: `AttributeError: 'float' object has no attribute 'lstrip'`
Uncomment the cell below to see the whole error output.

In [87]:
#hits = hits.apply(clean_hits)

There's an error when processing because NaN values are of type `float`. I will convert them to 0 and then process.

In [88]:
hits = hits.fillna('0')

In [89]:
hits = hits.apply(clean_hits)

Now convert the 0 values to NaN again, so they can be handled later.

In [90]:
hits[hits == 0] = np.nan

In [91]:
hits.isna().sum()

2595

Everything is ok now. Assign to dataframe column:

In [92]:
df.hits = hits

### After a first cleaning stage, let's see how things are now

In [93]:
# just take 5 rows to check the values and see if they effectively must be integers or some need a type conversion
string_vars = df.select_dtypes(object).sample(5)
string_vars.T

11859  \
name                                                    N. Kocik   
longname                                           Nicolas Kocik   
photourl        https://cdn.sofifa.com/players/232/417/21_60.png   
playerurl      http://sofifa.com/player/232417/nicolas-kocik/...   
nationality                                               France   
club                                     \n\n\n\nValenciennes FC   
contract                                             2016 ~ 2023   
positions                                                     GK   
height                                                     193cm   
best_position                                                 GK   
loan_date_end                                                NaN   
a/w                                                       Medium   
d/w                                                       Medium   

                                                           10530  \
name                                                    A. Doffo   
longname                                           Agustín Doffo   
photourl        https://cdn.sofifa.com/players/228/809/21_60.png   
playerurl      http://sofifa.com/player/228809/agustin-doffo/...   
nationality                                            Argentina   
club                                 \n\n\n\nClub Atlético Colón   
contract                                             2020 ~ 2020   
positions                                                CAM, RM   
height                                                     179cm   
best_position                                                CAM   
loan_date_end                                                NaN   
a/w                                                         High   
d/w                                                       Medium   

                                                           16170  \
name                                              H. Al Mansoury   
longname                                       Hamad Al Mansoury   
photourl        https://cdn.sofifa.com/players/247/066/21_60.png   
playerurl      http://sofifa.com/player/247066/hamad-al-manso...   
nationality                                 United Arab Emirates   
club                                                     No Club   
contract                                                    Free   
positions                                                     GK   
height                                                     178cm   
best_position                                                 GK   
loan_date_end                                                NaN   
a/w                                                       Medium   
d/w                                                       Medium   

                                                           3955   \
name                                                     D. Odoi   
longname                                              Denis Odoi   
photourl        https://cdn.sofifa.com/players/185/349/21_60.png   
playerurl      http://sofifa.com/player/185349/denis-odoi/210...   
nationality                                              Belgium   
club                                              \n\n\n\nFulham   
contract                                             2016 ~ 2022   
positions                                                 RB, CB   
height                                                     179cm   
best_position                                                 RB   
loan_date_end                                                NaN   
a/w                                                         High   
d/w                                                       Medium   

                                                           15771  
name                                                   M. Rundić  
longname                                            Milan Rundić  
photourl        https://cdn.sofifa.com/players

In a second round of cleaning, I will act on variables: `a/w` and `d/w`

### 9. Set `a/w` and `d/w` to categorical and sort their category values

In [94]:
df['a/w'].isna().sum()

0

In [95]:
df['d/w'].isna().sum()

0

In [96]:
df['a/w'] = df['a/w'].astype('category')
df['d/w'] = df['d/w'].astype('category')

In [97]:
df['a/w'].unique()

['Medium', 'High', 'Low']
Categories (3, object): ['High', 'Low', 'Medium']

In [98]:
df['d/w'].unique()

['Low', 'Medium', 'High']
Categories (3, object): ['High', 'Low', 'Medium']

In [99]:
df['a/w'] = df['a/w'].cat.set_categories(['Low', 'Medium', 'High'], ordered=True)
df['d/w'] = df['d/w'].cat.set_categories(['Low', 'Medium', 'High'], ordered=True)

In [100]:
df['a/w'].sort_values()

8606      Low
3451      Low
7347      Low
12569     Low
15007     Low
         ... 
3767     High
3768     High
14656    High
3758     High
8604     High
Name: a/w, Length: 18979, dtype: category
Categories (3, object): ['Low' < 'Medium' < 'High']

### 10. Clean `contract` variable

In [101]:
df.contract

0        2004 ~ 2021
1        2018 ~ 2022
2        2014 ~ 2023
3        2015 ~ 2023
4        2017 ~ 2022
            ...     
18974    2018 ~ 2022
18975    2020 ~ 2021
18976    2019 ~ 2020
18977    2020 ~ 2022
18978    2019 ~ 2023
Name: contract, Length: 18979, dtype: object

Let's see if there are any other "strange" values that don't start with a year

In [102]:
df.contract[ ~ df.contract.str.startswith('2')]

205      Jun 30, 2021 On Loan
248      Jun 30, 2021 On Loan
254      Jun 30, 2021 On Loan
289                      Free
292                      Free
                 ...         
18472    Aug 31, 2021 On Loan
18571    Jun 30, 2021 On Loan
18600    Dec 31, 2020 On Loan
18622    Dec 31, 2020 On Loan
18680    Dec 31, 2020 On Loan
Name: contract, Length: 1251, dtype: object

I will create a new column called `on_loan_from` for those players who are currently on loan.
For the rest of them, I will create two new columns: `contract_started` and `contract_ends`.  
`contract` column will be reasigned as a boolean column, meaning `free` has a value of False. Otherwise, `True` and will have his/her corresponding `contract_started` and `contract_ends` values.

Now I will create a custom function to deal with different types of contract values.

In [103]:
contract = df.contract.copy()

In [104]:
def process_contract(value):
    # contract has a start and end date
    if '~' in value:
        dates = value.split(' ~ ')
        contract_start = dates[0]
        contract_end = dates[1]
        return [contract_start, contract_end]
    # player has currently no contract
    elif value == 'Free':
        return ['free_player']
    # on loan contract
    else:
        loan_start = value[:-8]
        return ['on_loan', loan_start]      

In [105]:
df.contract.iloc[18472][:-8]

'Aug 31, 2021'

In [106]:
contract = contract.apply(process_contract)

After having applied the function, there will be three types of values:
1. a list with start and end years for contract
2. a list with only one element: 'free'
3. a list with first element equals to 'on_loan' and 2nd element: on_loan_from

Now let's create the three new columns: `contract_started`, `contract_ends`, `has_contract` (boolean) and `on_loan_from`

Create `on_loan` column:

In [107]:
df['on_loan_from'] = [i[1] if i[0] == 'on_loan' else np.nan for i in contract]

In [108]:
df['contract_started'] = [i[0] if len(i[0]) == 4 else np.nan for i in contract]

In [109]:
df['contract_ends'] = [i[1] if len(i[0]) == 4 else np.nan for i in contract]

In [110]:
df['has_contract'] = [bool(i) for i in df.contract_started]

In [111]:
df[['contract_started', 'contract_ends', 'has_contract', 'on_loan_from']]

contract_started contract_ends  has_contract on_loan_from
0                 2004          2021          True          NaN
1                 2018          2022          True          NaN
2                 2014          2023          True          NaN
3                 2015          2023          True          NaN
4                 2017          2022          True          NaN
...                ...           ...           ...          ...
18974             2018          2022          True          NaN
18975             2020          2021          True          NaN
18976             2019          2020          True          NaN
18977             2020          2022          True          NaN
18978             2019          2023          True          NaN

[18979 rows x 4 columns]

The last step is to convert to datetime these new columns

In [112]:
df.contract_started = pd.to_datetime(df.contract_started)
df.contract_ends = pd.to_datetime(df.contract_ends)
df.on_loan_from = df.on_loan_from = pd.to_datetime(df.on_loan_from)

### 11. Clean `club` variable

In [113]:
clubs = df.club.copy()
clubs

0               \n\n\n\nFC Barcelona
1                   \n\n\n\nJuventus
2            \n\n\n\nAtlético Madrid
3            \n\n\n\nManchester City
4        \n\n\n\nParis Saint-Germain
                    ...             
18974             \n\n\n\nWuhan Zall
18975        \n\n\n\nOldham Athletic
18976             \n\n\n\nDerry City
18977       \n\n\n\nDalian YiFang FC
18978       \n\n\n\nDalian YiFang FC
Name: club, Length: 18979, dtype: object

In [114]:
clubs = clubs.str.strip()
clubs.sample(5)

12580                CD Mirandés
8860           Sint-Truidense VV
18814         Wellington Phoenix
3127     Wolverhampton Wanderers
12733         Milton Keynes Dons
Name: club, dtype: object

In a previous execution of the cell above, I noticed that some club names have a number at the beginning. I will check and then remove the numbers.

In [115]:
clubs.isna().sum()

0

In [116]:
clubs[clubs.apply(lambda x: x[0].isnumeric() )]

355         1. FC Union Berlin
818         1. FC Union Berlin
967                 1. FC Köln
993                 1. FC Köln
1164                1. FC Köln
                 ...          
17576     1. FC Kaiserslautern
17874        1. FC Saarbrücken
18057    1. FC Heidenheim 1846
18134        1. FC Saarbrücken
18817     1. FC Kaiserslautern
Name: club, Length: 234, dtype: object

234 club names start with '1. '. I Will check if ALL OF THEM start with '1. '

In [117]:
clubs[clubs.str.startswith('1. ')].count()

234

Let's remove the string '1. ' from all of them.

In [118]:
clubs = clubs.str.lstrip('1. ')

In [119]:
clubs[clubs.str.startswith('1. ')].count()

0

All of them have the '1. ' string cleaned!

In [120]:
clubs.sample(20)

7132         Oud-Heverlee Leuven
8203                ADO Den Haag
9693                   Liverpool
13435     Nacional de Montevideo
15000                   Cádiz CF
10235              Coventry City
18952    Independiente del Valle
295              West Ham United
17465                 VfB Lübeck
5410                 Hammarby IF
8841       AC Mineros de Guayana
16760            Kashima Antlers
8797            Dalian YiFang FC
18119                   Al Fayha
1595                        SPAL
12200                      Parma
5330                   LA Galaxy
17639              Astra Giurgiu
14628                     Emelec
5607                Club Bolívar
Name: club, dtype: object

From a previous execution where  I took a sample of 20 different club values, a value of `No Club` arose. I will replace them by NaN values.

In [121]:
clubs[clubs == 'No Club'].count()

237

In [122]:
clubs[clubs == 'No Club'] = np.nan

Finally, set the cleaned club values to `club` variable in the dataset.

In [123]:
df.club = clubs

### 12. Categorizing more variables

In [124]:
df.select_dtypes(object).columns

Index(['name', 'longname', 'photourl', 'playerurl', 'nationality', 'club',
       'contract', 'positions', 'height', 'best_position', 'loan_date_end'],
      dtype='object')

Checking the object variables that could be turned into categories, I have just noticed I haven't processed the `loan_date_end`. Next step will be processing it. But first, let's categorize.

`positions` and `best_position` should be processed. As `positions` is a list of all the positions the player can play at, i will transform it to a python list and leave it as string. `best_position` will be converted to category dtype.

For everyone asking themselves why use `category` dtype, it's a matter of memory usage and optimisation. In datasets this size or even larger, memory can be an aspect to take into account.

In [125]:
positions = df.positions.copy()

In [126]:
positions.sample(10)

4256         CAM, CM
18090         ST, CF
5447         CDM, CM
8819              ST
9148              LB
16262    CDM, CB, RM
5894          RM, RB
5707              GK
8939         CM, CAM
7570     CAM, RM, ST
Name: positions, dtype: object

In [127]:
positions = positions.str.split(', ')

If someone asks himself / herself how to work with the positions data...  
Every row in the `positions` column has become a list. So you must access the list and inspect its content to look for an especific value. Below there's an example.

Get all the players that can play in LB (Left Full-Back; Block other team and hold a line)

In [128]:
# maybe not the finest solution, but I'm learning as most of you...
positions[positions.apply(lambda i : 'LB' in i)]

32                [LB]
60                [LB]
87            [LB, LM]
106               [LB]
108               [LB]
             ...      
18854    [LB, LWB, LW]
18871         [CB, LB]
18906             [LB]
18947         [LB, LW]
18978         [CB, LB]
Name: positions, Length: 2062, dtype: object

In [129]:
df.positions = positions

Now i will categorize the `best_position` variable. But first, some checking for weird values or nans.

In [130]:
best_pos = df.best_position.copy()

In [131]:
best_pos.dtype

dtype('O')

In [132]:
best_pos.unique()

array(['RW', 'ST', 'GK', 'CAM', 'LW', 'CB', 'CDM', 'CF', 'CM', 'RB', 'LB',
       'LM', 'RM', 'LWB', 'RWB'], dtype=object)

In [133]:
best_pos.isna().sum()

0

Finally, convert to categorical

In [134]:
df.best_position = best_pos.astype('category')

### 13. Clean `loan_end_date`

In [135]:
loan = df.loan_date_end.copy()

In [136]:
loan.isna().sum()

17966

In [137]:
loan.unique()

array([nan, 'Jun 30, 2021', 'Dec 31, 2020', 'Jan 30, 2021',
       'Jun 30, 2022', 'May 31, 2021', 'Jul 5, 2021', 'Dec 31, 2021',
       'Jul 1, 2021', 'Jan 1, 2021', 'Aug 31, 2021', 'Jan 31, 2021',
       'Dec 30, 2021', 'Jun 23, 2021', 'Jan 3, 2021', 'Nov 27, 2021',
       'Jan 17, 2021', 'Jun 30, 2023', 'Jul 31, 2021', 'Nov 22, 2020',
       'May 31, 2022', 'Dec 30, 2020', 'Jan 4, 2021', 'Nov 30, 2020',
       'Aug 1, 2021'], dtype=object)

This variable contains dates, so I will use `pd.to_datetime(variable)` to convert them to `datetime` dtype. Other date formats can be applied (eg. 'DD-MM-YYYY', european format)

In [138]:
loan = pd.to_datetime(loan)

Pandas has a special NaN value for dates that have a null value> NaT (Not a Time). I will show you some of them:

In [139]:
loan[loan.isnull()][:5]

0   NaT
1   NaT
2   NaT
3   NaT
4   NaT
Name: loan_date_end, dtype: datetime64[ns]

Check the rest of them to see if datetime conversion was ok

In [140]:
loan[loan.notnull()]

205     2021-06-30
248     2021-06-30
254     2021-06-30
302     2021-06-30
306     2021-06-30
           ...    
18472   2021-08-31
18571   2021-06-30
18600   2020-12-31
18622   2020-12-31
18680   2020-12-31
Name: loan_date_end, Length: 1013, dtype: datetime64[ns]

In [141]:
df.loan_date_end = loan

### 14. Reestructuration and/or removal of columns after all processing

In [142]:
df.columns

Index(['id', 'name', 'longname', 'photourl', 'playerurl', 'nationality', 'age',
       '↓ova', 'pot', 'club', 'contract', 'positions', 'height', 'weight',
       'preferred_foot', 'bov', 'best_position', 'joined', 'loan_date_end',
       'value_euros_M', 'annual_wage_euros_M', 'release_clause_euros_M',
       'attacking', 'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'skill', 'dribbling', 'curve',
       'fk_accuracy', 'long_passing', 'ball_control', 'movement',
       'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance',
       'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'mentality', 'aggression', 'interceptions', 'positioning', 'vision',
       'penalties', 'composure', 'defending', 'marking', 'standing_tackle',
       'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats',
       'base_stats', 'w/f', 'sm', 'a/w', 'd/w', 'ir', 'pac

I will remove the following columns: `contract` and `height`.  
In columns `annual_wage_euros_M`, `release_euros_M` and `value_euros_M`, the string 'euros_M' will be replaced by 'million_euros'.  
Column `weight` will be renamed to `weight_kg`

In [143]:
df.drop(['contract', 'height'], axis=1, inplace=True)

In [144]:
df = df.rename(columns = {'annual_wage_euros_M' : 'annual_wage_million_euros',
                          'release_clause_euros_M' : 'release_clause_million_euros',
                          'value_euros_M' : 'value_million_euros',
                         'weight' : 'weight_kg'})

In [145]:
df.columns[df.columns.str.contains('euros')]

Index(['value_million_euros', 'annual_wage_million_euros',
       'release_clause_million_euros'],
      dtype='object')

# Study of the numeric variables

In [146]:
num_vars = df.select_dtypes('number').columns.to_list()
print(num_vars)

['id', 'age', '↓ova', 'pot', 'weight_kg', 'bov', 'value_million_euros', 'annual_wage_million_euros', 'release_clause_million_euros', 'attacking', 'crossing', 'finishing', 'heading_accuracy', 'short_passing', 'volleys', 'skill', 'dribbling', 'curve', 'fk_accuracy', 'long_passing', 'ball_control', 'movement', 'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance', 'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots', 'mentality', 'aggression', 'interceptions', 'positioning', 'vision', 'penalties', 'composure', 'defending', 'marking', 'standing_tackle', 'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats', 'base_stats', 'pac', 'sho', 'pas', 'dri', 'def', 'phy', 'hits', 'height_m', 'height_feet']


In [147]:
df[num_vars].dtypes

id               int64
age              int64
↓ova             int64
pot              int64
weight_kg      float64
                ...   
def              int64
phy              int64
hits           float64
height_m       float64
height_feet    float64
Length: 61, dtype: object

## Study of missing values

In [148]:
def nan_percent(series):
    nan_count = series.isna().sum()
    total_count = series.shape[0]
    return round(nan_count / total_count * 100, 2)

In [149]:
missing_values_pctg = df[num_vars].apply(nan_percent, axis=0).sort_values(ascending=False)
missing_values_pctg.name = 'percentage of missing values'
missing_values_pctg

hits           13.67
id              0.00
gk_diving       0.00
mentality       0.00
aggression      0.00
               ...  
reactions       0.00
balance         0.00
power           0.00
shot_power      0.00
height_feet     0.00
Name: percentage of missing values, Length: 61, dtype: float64

Looks like there are no nan values (except for `hits` variable). That doesn't mean that there are no invalid values. It requires a deeper exploration.  
I will get a sample of the numeric variables, in order to make a visual inspection for strange values.

In [150]:
pd.set_option('display.max_rows', 100)

In [151]:
df[num_vars].sample(10).T

12180          15901          12375  \
id                            244800.000000  234733.000000  213110.000000   
age                               24.000000      22.000000      28.000000   
↓ova                              63.000000      59.000000      63.000000   
pot                               68.000000      69.000000      63.000000   
weight_kg                         64.000000      76.000000      84.000000   
bov                               64.000000      60.000000      63.000000   
value_million_euros                0.675000       0.500000       0.500000   
annual_wage_million_euros          0.052000       0.104000       0.052000   
release_clause_million_euros       0.563000       0.488000       0.656000   
attacking                        238.000000     243.000000     287.000000   
crossing                          58.000000      51.000000      42.000000   
finishing                         42.000000      57.000000      62.000000   
heading_accuracy                  41.000000      44.000000      67.000000   
short_passing                     61.000000      51.000000      59.000000   
volleys                           36.000000      40.000000      57.000000   
skill                            269.000000     240.000000     268.000000   
dribbling                         62.000000      62.000000      63.000000   
curve                             58.000000      52.000000      59.000000   
fk_accuracy                       32.000000      23.000000      41.000000   
long_passing                      56.000000      46.000000      42.000000   
ball_control                      61.000000      57.000000      63.000000   
movement                         386.000000     358.000000     299.000000   
acceleration                      88.000000      84.000000      65.000000   
sprint_speed                      83.000000      80.000000      72.000000   
agility                           83.000000      76.000000      70.000000   
reactions                         56.000000      47.000000      45.000000   
balance                           76.000000      71.000000      47.000000   
power                            282.000000     281.000000     352.000000   
shot_power                        57.000000      49.000000      68.000000   
jumping                           57.000000      54.000000      74.000000   
stamina                           69.000000      63.000000      73.000000   
strength                          48.000000      59.000000      75.000000   
long_shots                        51.000000      56.000000      62.000000   
mentality                        274.000000     208.000000     241.000000   
aggression                        63.000000      28.000000      49.000000   
interceptions                     52.000000      32.000000      22.000000   
positioning                       62.000000      56.000000      60.000000   
vision                            52.000000      53.000000      43.000000   
penalties                         45.000000      39.000000      67.000000   
composure                         60.000000      53.000000      56.000000   
defending                        177.000000      90.000000      69.000000   
marking                           62.000000      30.000000      27.000000   
standing_tackle                   58.000000      30.000000      22.000000   
sliding_tackle                    57.000000      30.000000      20.000000   
goalkeeping                       40.000000      50.000000      54.000000   
gk_diving                          8.000000       5.000000      11.000000   
gk_handling                        8.000000       8.000000       7.000000   
gk_kicking                         8.000000      14.000000      16.000000   
gk_positioning                     7.000000      14.000000      14.000000   
gk_reflexes                        9.000000       9.000000       6.000000   
total_stats                     1666.000000    1470.000000    1570.000000   
base_stats                       3

The NaN values in `hits` variable may mean that certain profile websites have never been visited. I will fill them with 0 visits.

In [152]:
hits.isna().sum()

2595

In [153]:
hits = hits.fillna(0)

In [154]:
hits[hits == 0].count()

2595

All NaN values from `hits` have been successfully filled with 0 visits values.

## Dealing with outliers

There are too may numeric variables, so I will create a function to check if a variable has outliers.  
A value is considered an outliers if it lies outside the limits of +/- 1.5 · IR (Interquartile Range). 

In [155]:
def has_outliers(series):
    Q1, Q3 = series.quantile([.25, .75])
    IR = Q3 - Q1
    below_Q1 = series < (Q1 - 1.5*IR)
    above_Q3 = series > (Q3 + 1.5*IR)
    number_outliers = (below_Q1 | above_Q3).sum()
    if number_outliers:
        return number_outliers

In [156]:
df[num_vars]

id  age  ↓ova  pot  weight_kg  bov  value_million_euros  \
0      158023   33    93   93       72.0   93               103.50   
1       20801   35    92   92       83.0   92                63.00   
2      200389   27    91   93       87.0   91               120.00   
3      192985   29    91   91       70.0   91               129.00   
4      190871   28    91   91       68.0   91               132.00   
...       ...  ...   ...  ...        ...  ...                  ...   
18974  247223   21    47   55       66.0   49                 0.10   
18975  258760   17    47   67       65.0   51                 0.13   
18976  252757   18    47   65       74.0   49                 0.12   
18977  243790   20    47   57       69.0   48                 0.10   
18978  252520   21    47   57       75.0   50                 0.10   

       annual_wage_million_euros  release_clause_million_euros  attacking  \
0                        29.1200                       138.400        429   
1                        11.4400                        75.900        437   
2                         6.5000                       159.400         95   
3                        19.2400                       161.000        407   
4                        14.0400                       166.500        408   
...                          ...                           ...        ...   
18974                     0.0520                         0.070        145   
18975                     0.0026                         0.165        211   
18976                     0.0026                         0.131        200   
18977                     0.1040                         0.088        215   
18978                     0.0520                         0.079        163   

       crossing  finishing  heading_accuracy  short_passing  volleys  skill  \
0            85         95                70             91       88    470   
1            84         95                90             82       86    414   
2            13         11                15             43       13    109   
3            94         82                55             94       82    441   
4            85         87                62             87       87    448   
...         ...        ...               ...            ...      ...    ...   
18974        23         26                43             26       27    142   
18975        38         42                40             56       35    219   
18976        30         34                43             54       39    207   
18977        45         52                34             42       42    194   
18978        40         18                40             45       20    171   

       dribbling  curve  fk_accuracy  long_passing  ball_control  movement  \
0             96     93           94            91            96       451   
1             88     81           76            77            92       431   
2             12     13           14            40            30       307   
3             88     85           83            93            92       398   
4             95     88           89            81            95       453   
...          ...    ...          ...           ...           ...       ...   
18974         27     23           21            29            42       294   
18975         46     40           35            50            48       305   
18976         43     39           31            47            47       290   
18977         51     35           31            31            46       254   
18978         40     33           20            40            38       279   

       acceleration  sprint_speed  agility  reactions  balance  power  \
0                91            80       91         94       95    389   
1                87            91       87         95       71    444   
2                43            60       67         88       49    268   
3                77            76       78         91       76    4

In [157]:
df[num_vars].apply(has_outliers, axis=0).T.sort_values(ascending=False)[:15]

hits                            2202.0
value_million_euros             2199.0
goalkeeping                     2181.0
gk_kicking                      2089.0
gk_positioning                  2086.0
gk_reflexes                     2085.0
gk_handling                     2085.0
gk_diving                       2085.0
ball_control                    2053.0
release_clause_million_euros    2020.0
attacking                       2016.0
annual_wage_million_euros       1999.0
dribbling                       1962.0
short_passing                   1748.0
mentality                       1493.0
dtype: float64

There seems to be a huge range of values for wages, value and hits. Outliers in `hit` column could be due to big differences in popularity (famous players get huge amounts of clicks in their websites, compared to almost none for unknown players). The same for wage and value: wide difference between Messi and a relatively unknown player of a modest football club.

I see many outliers in the goalkeeper variables. Not all players are goalkeepers, so maybe a good idea would be to set NaN at those variables for non-goalkeeper players.

### Soccer / football position abbreviations

Abbreviations for Soccer Midfielder Positions
- CB: Center-back; Stop other team from getting near GK
- GK: Goal Keeper; Stops ball from going in goal
- LB: Left Full-Back; Block other team and hold a line
- LWB: Left Wing-Back; Play near sidelines and block other team
- RB: Right Full-Back; Block other team and hold a line
- RWB: Right Wing-Back; Play near sidelines and block other team
- SW: Sweeper; Stands between GK and everyone else as last line of defense

Midfielders need to be good at both offense and defense because they typically play toward the middle of the field and near their opponent’s goal.

- AM: Attacking Midfielder; Focuses on shooting and scoring
- CM: Center Midfielder; Links defense and offense of a team
- DM: Defensive Midfielder; Organize attacking plays
- LW: Left Winger; Plays along sideline to shoot, assist, and score
- RW: Right Winger; Plays along sideline to shoot, assist, and score

Abbreviations for Soccer Forward Positions

Forwards are the players who play closer to their opponent’s goal and try to score goals.

- CF: Center Forward; Link offense and defense of a team
- S: Striker; Focuses on scoring

In [158]:
goalkeeper_attributes = [i for i in df.columns if i.startswith('gk') ]
df[['positions'] + goalkeeper_attributes]

positions  gk_diving  gk_handling  gk_kicking  gk_positioning  \
0      [RW, ST, CF]          6           11          15              14   
1          [ST, LW]          7           11          15              14   
2              [GK]         87           92          78              90   
3         [CAM, CM]         15           13           5              10   
4         [LW, CAM]          9            9          15              15   
...             ...        ...          ...         ...             ...   
18974          [CB]          7            8           5              14   
18975          [CM]         12           10           9               6   
18976          [CM]         13           12           6               6   
18977          [RW]         14           12           9               8   
18978      [CB, LB]         10           11           7              10   

       gk_reflexes  
0                8  
1               11  
2               90  
3               13  
4               11  
...            ...  
18974           11  
18975            8  
18976           11  
18977           12  
18978           12  

[18979 rows x 6 columns]

In [159]:
df[goalkeeper_attributes][df.positions.apply(lambda x : 'GK' in x)].mean()

gk_diving         65.245783
gk_handling       62.944578
gk_kicking        61.772048
gk_positioning    63.296867
gk_reflexes       66.186506
dtype: float64

Watching the mean values of goalkeeper's attributes, one can see that the presence of outliers in those variables is due to most players not being a goalkeeper. The average values for goalkeepers are higher than those for non-goalkeeper players.

One last step about analyzing outliers: the percentage of outliers. I will take into acount only those variables that have a significative % (% >= 5) of outliers. This minimum threshold can be easily changed in the code below.

In [160]:
outliers_pctg = ( df[num_vars].apply(has_outliers, axis=0) / df.shape[0] * 100 ).dropna()
outliers_pctg = round( outliers_pctg, 2).sort_values(ascending=False)
outliers_pctg[outliers_pctg >= 5]

hits                            11.60
value_million_euros             11.59
goalkeeping                     11.49
gk_kicking                      11.01
gk_reflexes                     10.99
gk_positioning                  10.99
gk_handling                     10.99
gk_diving                       10.99
ball_control                    10.82
release_clause_million_euros    10.64
attacking                       10.62
annual_wage_million_euros       10.53
dribbling                       10.34
short_passing                    9.21
mentality                        7.87
skill                            6.50
heading_accuracy                 5.84
dtype: float64

## Renaming some numeric variables

Some of the numeric variables have a non-descriptive name, so making use of the dataset dictionary, I will give them a more self-explanatory name.  
At the beginning of the dataset i created a dataframe (`var_description_df`) with the description of the variables. I will use it now to help me with the renaming.

In [161]:
var_description_df.head()

description
variable                                                
photoUrl                  The URL of the player's photo.
LongName                    The full name of the player.
playerUrl    The URL of the player's page on sofifa.com.
Nationality               The nationality of the player.
Positions             The positions the player can play.

Variable names with only about 3 or 4 letters need some more explanatory names. Let's figure out which of them, making a pandas query...

In [162]:
short_var_names = [i for i in df.columns if len(i) <=4]
print(short_var_names, sep=', ')

['id', 'name', 'age', '↓ova', 'pot', 'club', 'bov', 'w/f', 'sm', 'a/w', 'd/w', 'ir', 'pac', 'sho', 'pas', 'dri', 'def', 'phy', 'hits']


Now I will retrieve the variable names and their description from the dataframe

#### Error: `KeyError: "None of [Index(['id', 'name', 'age', '↓ova', 'pot', 'club', 'bov', 'w/f', 'sm', 'a/w',\n       'd/w', 'ir', 'pac', 'sho', 'pas', 'dri', 'def', 'phy', 'hits'],\n      dtype='object')] are in the [columns]"`
Uncomment the cell below to see the whole error output.

In [163]:
#var_description_df[short_var_names]

The above error was caused because the `var_description_df` had the *old* names, before they were processed to replace spaces by underscore (_) and set to lowercase.  
I will update the dataframe to reflect the changes.

In [164]:
old_var_names = var_description_df.index
old_var_names

Index(['photoUrl', 'LongName', 'playerUrl', 'Nationality', 'Positions', 'Name',
       'Age', 'OVA', 'POT', 'Team & Contract', 'ID', 'Height', 'Weight',
       'foot', 'BOV', 'BP', 'Growth', 'Joined', 'Loan Date End', 'Value',
       'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 'PHY', 'Hits'

In [165]:
current_var_names = df.columns
current_var_names

Index(['id', 'name', 'longname', 'photourl', 'playerurl', 'nationality', 'age',
       '↓ova', 'pot', 'club', 'positions', 'weight_kg', 'preferred_foot',
       'bov', 'best_position', 'joined', 'loan_date_end',
       'value_million_euros', 'annual_wage_million_euros',
       'release_clause_million_euros', 'attacking', 'crossing', 'finishing',
       'heading_accuracy', 'short_passing', 'volleys', 'skill', 'dribbling',
       'curve', 'fk_accuracy', 'long_passing', 'ball_control', 'movement',
       'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance',
       'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'mentality', 'aggression', 'interceptions', 'positioning', 'vision',
       'penalties', 'composure', 'defending', 'marking', 'standing_tackle',
       'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats',
       'base_stats', 'w/f', 'sm', 'a/w', 'd/w', 'ir', 'pac'

Some of the old variables have disappeared, so the description dataframe needs some processing. I will take out the variables that have remained the same and later I will process the ones that have changed.

In [166]:
var_descrip2 = var_description_df.copy()

Now the cleaning of the variables must be done to this copy of the description dataframe (refer to Proposed Actions, 1. Rename columns)

In [167]:
var_descrip2_backup = var_descrip2.columns.to_list()
var_descrip2.index = [i.replace(' & ', '_').replace(' ', '_').lower() for i in var_descrip2.index]

In [168]:
var_descrip2.sample(10)

description
age                                        The age of the player.
foot                            The preferred foot of the player.
name                                The short name of the player.
hits            This refers to the number of times the player'...
growth          The difference between the potential rating an...
sm              This refers to the player's skill moves ability. 
gk_positioning  The goalkeeper positioning attribute of the pl...
ova                  The overall rating of the player in FIFA 21.
positions                      The positions the player can play.
id                          The unique identifier for the player.

In [169]:
var_descrip2.index

Index(['photourl', 'longname', 'playerurl', 'nationality', 'positions', 'name',
       'age', 'ova', 'pot', 'team_contract', 'id', 'height', 'weight', 'foot',
       'bov', 'bp', 'growth', 'joined', 'loan_date_end', 'value', 'wage',
       'release_clause', 'attacking', 'crossing', 'finishing',
       'heading_accuracy', 'short_passing', 'volleys', 'skill', 'dribbling',
       'curve', 'fk_accuracy', 'long_passing', 'ball_control', 'movement',
       'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance',
       'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'mentality', 'aggression', 'interceptions', 'positioning', 'vision',
       'penalties', 'composure', 'defending', 'marking', 'standing_tackle',
       'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats',
       'base_stats', 'w/f', 'sm', 'a/w', 'd/w', 'ir', 'pac', 'sho', 'pas',
       'dri', 'def', 'phy', 'hits'],

`unchanged_vars` are the ones who are in the clean dataframe and also have a description in `var_description_df`. I have noticed that some variables from the fifa21 version 2 are not in the descriptions.

In [170]:
unchanged_vars = [i for i in var_descrip2.index if i in df.columns]
print(unchanged_vars, sep=', ')

['photourl', 'longname', 'playerurl', 'nationality', 'positions', 'name', 'age', 'pot', 'id', 'bov', 'joined', 'loan_date_end', 'attacking', 'crossing', 'finishing', 'heading_accuracy', 'short_passing', 'volleys', 'skill', 'dribbling', 'curve', 'fk_accuracy', 'long_passing', 'ball_control', 'movement', 'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance', 'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots', 'mentality', 'aggression', 'interceptions', 'positioning', 'vision', 'penalties', 'composure', 'defending', 'marking', 'standing_tackle', 'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats', 'base_stats', 'w/f', 'sm', 'a/w', 'd/w', 'ir', 'pac', 'sho', 'pas', 'dri', 'def', 'phy', 'hits']


In [171]:
changed_vars = [i for i in var_descrip2.index if i not in df.columns]
print(changed_vars, sep=', ')

['ova', 'team_contract', 'height', 'weight', 'foot', 'bp', 'growth', 'value', 'wage', 'release_clause']


Remove the changed vars in the description dataframe. They will be added manually (as they are only a few) with a description.

In [172]:
# the var_descript2 dataframe keeps (by now) only the unchanged variables
var_descrip2 = var_descrip2.loc[unchanged_vars]
var_descrip2.sample(5)

description
name                               The short name of the player.
loan_date_end          The date the player's loan contract ends.
gk_reflexes    This refers to the goalkeeper's ability to rea...
positioning             The positioning attribute of the player.
pac            This refers to the player's pace or speed attr...

Maybe some of them don't belong to the clean dataframe anymore, not even in a renamed version. I will create a function that correlates the old variable name and its new renamed variables.

In [173]:
columns = df.columns
var_to_be_removed = []
for var in changed_vars:
    new_cols = columns[columns.str.contains(var)]
    if new_cols.any():
        print(f'Old name: {var} --> new name: {new_cols.values}')
    else:
        var_to_be_removed.append(var)
print('\n\nThese variables need to be removed from variable description dataframe:')
print(var_to_be_removed)

Old name: ova --> new name: ['↓ova']
Old name: height --> new name: ['height_m' 'height_feet']
Old name: weight --> new name: ['weight_kg']
Old name: foot --> new name: ['preferred_foot']
Old name: value --> new name: ['value_million_euros']
Old name: wage --> new name: ['annual_wage_million_euros']
Old name: release_clause --> new name: ['release_clause_million_euros']


These variables need to be removed from variable description dataframe:
['team_contract', 'bp', 'growth']


### Process manually the changed variables

#### `ova` var.  
This needs to be cleaned in the clean dataframe to remove the downside arrow in the var name (copy and paste to get the arrow character)

In [174]:
df.columns[df.columns.str.endswith('ova')]

Index(['↓ova'], dtype='object')

In [175]:
df.rename(columns = { '↓ova' : 'ova'}, inplace=True)

In [176]:
df.columns[df.columns.str.endswith('ova')]

Index(['ova'], dtype='object')

#### In order to add rows to `var_descrip2` dataframe I will reset the index

In [177]:
var_descrip2.reset_index(inplace=True)

In [178]:
var_descrip2.rename(columns = {'index' : 'variable'}, inplace=True)

In [179]:
var_descrip2.head()

variable                                  description
0     photourl               The URL of the player's photo.
1     longname                 The full name of the player.
2    playerurl  The URL of the player's page on sofifa.com.
3  nationality               The nationality of the player.
4    positions           The positions the player can play.

#### Get and process the description of the variables to be added

In [180]:
ova_descript = var_description_df.loc['OVA'].values[0]
height_m_descript = var_description_df.loc['Height'].values[0].replace('feet and inches', 'metres')
height_feet_descript = var_description_df.loc['Height'].values[0].replace('feet and inches', 'feet')
weight_descript = var_description_df.loc['Weight'].values[0].replace('pounds', 'kg')
foot_descript = var_description_df.loc['foot'].values[0]
value_descript = var_description_df.loc['Value'].values[0][:-1] + ', in millions of euros.'
wage_descript = var_description_df.loc['Wage'].values[0][:-1] + ', in millions of euros.'
release_clause_descript = var_description_df.loc['Release Clause'].values[0][:-1] + ', in millions of euros.'

In [181]:
new_entries = {'ova' : ova_descript,
               'height_m' : height_m_descript,
               'height_ft' : height_feet_descript,
               'weight_kg' : weight_descript,
               'preferred_foot' : foot_descript,
               'value_million_euros' : value_descript,
               'annual_wage_million_euros' : wage_descript,
               'release_clause_million_euros' : release_clause_descript
              }

In [182]:
new_description = pd.DataFrame({'variable': new_entries.keys(),
                                'description' : new_entries.values()
                               })
new_description

variable  \
0                           ova   
1                      height_m   
2                     height_ft   
3                     weight_kg   
4                preferred_foot   
5           value_million_euros   
6     annual_wage_million_euros   
7  release_clause_million_euros   

                                         description  
0       The overall rating of the player in FIFA 21.  
1                The height of the player in metres.  
2                  The height of the player in feet.  
3                    The weight of the player in kg.  
4                  The preferred foot of the player.  
5  The market value of the player in FIFA 21, in ...  
6  The weekly wage of the player in FIFA 21, in m...  
7  The release clause value of the player in FIFA...

### Final steps with description dataframe
1. Concat the var dataframe and the new entries
2. Replace the processed description dataset by the original one
3. Export to csv.

In [183]:
var_description_df = pd.concat([var_descrip2, new_description], axis=0)
var_description_df.set_index('variable', inplace=True)

In [184]:
var_description_df.tail(10)

description
variable                                                                       
phy                           This refers to the player's physicality or str...
hits                          This refers to the number of times the player'...
ova                                The overall rating of the player in FIFA 21.
height_m                                    The height of the player in metres.
height_ft                                     The height of the player in feet.
weight_kg                                       The weight of the player in kg.
preferred_foot                                The preferred foot of the player.
value_million_euros           The market value of the player in FIFA 21, in ...
annual_wage_million_euros     The weekly wage of the player in FIFA 21, in m...
release_clause_million_euros  The release clause value of the player in FIFA...

A bit of preparation before exportation to csv: reset index and add it as the variable column

In [185]:
var_descript_csv = var_description_df.reset_index()
var_descript_csv = var_descript_csv.rename(columns = {'index' : 'variable'})

In [186]:
var_descript_csv.to_csv('variable_descriptions.csv', index=False)

### Rename the short named variables

In [187]:
short_var_names

['id',
 'name',
 'age',
 '↓ova',
 'pot',
 'club',
 'bov',
 'w/f',
 'sm',
 'a/w',
 'd/w',
 'ir',
 'pac',
 'sho',
 'pas',
 'dri',
 'def',
 'phy',
 'hits']

I will remove `id`, `name`, `age`, and `club` from list of variables to be renamed. I have just thought `hits` can be renamed to `profile_web_visits`.

In [188]:
short_var_names = short_var_names[3:]
short_var_names

['↓ova',
 'pot',
 'club',
 'bov',
 'w/f',
 'sm',
 'a/w',
 'd/w',
 'ir',
 'pac',
 'sho',
 'pas',
 'dri',
 'def',
 'phy',
 'hits']

#### Error: `KeyError: '↓ova'"`
Uncomment the cell below to see the whole error output.

In [189]:
#for var in short_var_names:
#    print(f'Var: {var}. Description: {var_description_df.loc[var]}')

In [190]:
short_var_names[0] = 'ova'

In [191]:
short_var_names.remove('club')

In [192]:
for var in short_var_names:
    print(f'Var: {var} -  Description: {var_description_df.loc[var].values[0]}')

Var: ova -  Description: The overall rating of the player in FIFA 21.
Var: pot -  Description: The potential rating of the player in FIFA 21.
Var: bov -  Description: The best overall rating the player has achieved in their career.
Var: w/f -  Description: This refers to the player's weaker foot ability.
Var: sm -  Description: This refers to the player's skill moves ability. 
Var: a/w -  Description: This refers to the player's attacking work rate. It measures how frequently the player participates in attacking actions, such as making runs or positioning themselves in the opponent's half.
Var: d/w -  Description: This refers to the player's defensive work rate. It measures how frequently the player participates in defensive actions, such as tracking back or making tackles.
Var: ir -  Description: This refers to the player's injury resistance. It measures the player's ability to avoid injuries and how quickly they recover from them.
Var: pac -  Description: This refers to the player's 

With the descriptions, I have the information to rename to a more descriptive way the variables:

In [193]:
renaming_dict = {
    'ova': 'overall_rating',
    'pot' : 'potential_rating',
    'bov' : 'best_overall_rating',
    'w/f' : 'weaker_foot_ability',
    'sm' : 'skill_moves_ability',
    'a/w' : 'attack_work_rate',
    'd/w' : 'defense_work_rate',
    'ir' : 'injury_resistance',
    'pac' : 'player_speed',
    'sho' : 'shooting_ability',
    'pas' : 'passing_ability',
    'dri' : 'dribbling_ability',
    'def' : 'defensive_ability',
    'phy' : 'player_strength',
    'hits' : 'profile_web_visits',
    'photourl' : 'photo_url',
    'playerurl' : 'player_profile_web',
}    

In [194]:
df.rename(columns = renaming_dict, inplace=True)

In [195]:
df.columns

Index(['id', 'name', 'longname', 'photo_url', 'player_profile_web',
       'nationality', 'age', 'overall_rating', 'potential_rating', 'club',
       'positions', 'weight_kg', 'preferred_foot', 'best_overall_rating',
       'best_position', 'joined', 'loan_date_end', 'value_million_euros',
       'annual_wage_million_euros', 'release_clause_million_euros',
       'attacking', 'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'skill', 'dribbling', 'curve',
       'fk_accuracy', 'long_passing', 'ball_control', 'movement',
       'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance',
       'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'mentality', 'aggression', 'interceptions', 'positioning', 'vision',
       'penalties', 'composure', 'defending', 'marking', 'standing_tackle',
       'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats',
   

# TODO (after renaming of columns the variable description dataset should be updated)

## Final review of whole dataset

In [196]:
df.sample(5).T

12880  \
id                                                                       210898   
name                                                                A. Al Fahad   
longname                                                      Abdullah Al Fahad   
photo_url                      https://cdn.sofifa.com/players/210/898/21_60.png   
player_profile_web            http://sofifa.com/player/210898/abdullah-al-fa...   
nationality                                                        Saudi Arabia   
age                                                                          26   
overall_rating                                                               63   
potential_rating                                                             66   
club                                                                    Al Raed   
positions                                                              [CB, RB]   
weight_kg                                                                  73.0   
preferred_foot                                                            Right   
best_overall_rating                                                          65   
best_position                                                                CB   
joined                                                      2018-06-09 00:00:00   
loan_date_end                                                               NaT   
value_million_euros                                                        0.55   
annual_wage_million_euros                                                 0.208   
release_clause_million_euros                                               0.69   
attacking                                                                   248   
crossing                                                                     38   
finishing                                                                    54   
heading_accuracy                                                             63   
short_passing                                                                57   
volleys                                                                      36   
skill                                                                       222   
dribbling                                                                    43   
curve                                                                        41   
fk_accuracy                                                                  33   
long_passing                                                                 58   
ball_control                                                                 47   
movement                                                                    331   
acceleration                                                                 65   
sprint_speed                                                                 69   
agility                                                                      72   
reactions                                                                    61   
balance                                                                      64   
power                                                                       343   
shot_power                                                                   53   
jumping                                                                      82   
stamina                                                                      76   
strength                                                                     80   
long_shots                                                                   52   
mentality                                                                   259   
aggression                                                                   58   
interceptions                                                                64   
positioning                                                                  39   
vision 

Watching the results, most of the variable with more or less high percentages are due to the before mentioned reasons, so I won't take action regarding to outliers.

In [197]:
df.to_csv('clean_fifa21_dataset.csv', index=False, na_rep='NULL')